ingest/expression_files/mtx.py

```diff
@@ -24,144 +20,221 @@
 
-        mtx_ingest_file = IngestFiles(mtx_path, self.ALLOWED_FILE_TYPES)
-        self.mtx_file, self.mtx_local_path = mtx_ingest_file.resolve_path(mtx_path)
         # A list ['N', 'K', 'M'] that represents a gene-barcode matrix where N
-        # is the gene index, M is the barcode index, and K is the expresion score
-        # for the given gene index
-        self.mtx_description = linecache.getline(self.mtx_local_path, 2).split()
+        # is the gene index, M is the barcode index, and K is the expression
+        # score for the given gene index
+        self.mtx_dimensions: List[int] = MTXIngestor.get_mtx_dimensions(self.mtx_file)
+
+    @staticmethod
+    def check_valid(barcodes: List[str], genes: List[str], mtx_dimensions):
+        error_messages = []
+
+        try:
+            MTXIngestor.check_bundle(barcodes, genes, mtx_dimensions)
+        except ValueError as v:
+            error_messages.append(str(v))
+        try:
+            MTXIngestor.check_duplicate_genes(genes)
+        except ValueError as v:
+            error_messages.append(str(v))
+        try:
+            GeneExpression.check_unique_cells(barcodes)
+        except ValueError as v:
+            error_messages.append(str(v))
+
+        if len(error_messages) > 0:
+            raise ValueError("; ".join(error_messages))
+        return True
+
+    @staticmethod
+    def check_bundle(barcodes, genes, mtx_dimensions):
+        """Confirms barcode and gene files have expected length of values"""
+        expected_genes = mtx_dimensions[0]
+        actual_genes = len(genes)
+
+        expected_barcodes = mtx_dimensions[1]
+        actual_barcodes = len(barcodes)
+
+        if (actual_barcodes == expected_barcodes) and (actual_genes == expected_genes):
+
+            return True
+        else:
+            msg = (
+                f"Expected {expected_barcodes} cells and {expected_genes} genes. "
+                f"Got {actual_barcodes} cells and {actual_genes} genes."
+            )
+            raise ValueError(msg)
+
+    @staticmethod
+    def check_duplicate_genes(gene_names: List):
+        unique_gene_names: List[str] = set(gene_names)
+        if len(unique_gene_names) != len(gene_names):
+            amount_of_duplicates = len(unique_gene_names) - len(gene_names)
+            msg = (
+                "Duplicate header values are not allowed."
+                f"There are {amount_of_duplicates} duplicates"
+            )
+            raise ValueError(msg)
+        return True
+
+    @staticmethod
+    def get_mtx_dimensions(file_handler) -> List:
+        for line in file_handler:
+            if not line.startswith("%"):
+                mtx_dimensions: List[str] = line.strip().split()
+                # Convert values in mtx_dimensions to int
+                return list(map(int, mtx_dimensions))
+
+    @staticmethod
+    def is_sorted(idx: int, visited_expression_idx: List[int]):
```

Comment:
I noticed all cases of `is_foo()` renamed to `check_foo()` in a recent PR.  For consistency, we should use one naming pattern or the other, but not both.

This would be consistent those recent changes:

```suggestion
    def check_sorted(idx: int, visited_expression_idx: List[int]):
```

Reply:
This is a check that happens in transform and not in check_valid(). For formatting issues, the rational was that we wanted to know all validation errors prior to ingest which is why "is" changed to "check".  If is_sorted() returns false parsing stops and immediately exits.

Reply:
Gotcha, helpful to know.  That rationale seems reasonable.